src/arc.rs

```diff
@@ -0,0 +1,215 @@
+    }
+}
+
+impl<T> Arc<T> {
+    pub fn new(inner: T) -> Arc<T> {
+        let bx = Box::new(ArcInner { inner, rc: AtomicUsize::new(1) });
+        let ptr = Box::into_raw(bx);
+        Arc { ptr }
+    }
+
+    pub fn copy_from_slice(s: &[T]) -> Arc<[T]> {
+        let align =
+            std::cmp::max(mem::align_of::<T>(), mem::align_of::<AtomicUsize>());
+
+        let rc_width = std::cmp::max(align, mem::size_of::<AtomicUsize>());
+        let data_width = mem::size_of::<T>().checked_add(s.len()).unwrap();
+
+        let size = rc_width.checked_add(data_width).unwrap();
+
+        let layout = Layout::from_size_align(size, align).unwrap();
+
+        unsafe {
+            let ptr = alloc(layout);
+
+            assert!(!ptr.is_null(), "failed to allocate Arc");
+            #[allow(clippy::cast_ptr_alignment)]
+            ptr::write(ptr as _, AtomicUsize::new(1));
+
+            let data_ptr = ptr.add(rc_width);
+            ptr::copy_nonoverlapping(s.as_ptr(), data_ptr as _, s.len());
+
+            let fat_ptr: *const ArcInner<[T]> = Arc::fatten(ptr, s.len());
+
+            Arc { ptr: fat_ptr as *mut _ }
+        }
+    }
+
+    /// <https://users.rust-lang.org/t/construct-fat-pointer-to-struct/29198/9>
+    #[allow(trivial_casts)]
+    fn fatten(data: *const u8, len: usize) -> *const ArcInner<[T]> {
+        // Requirements of slice::from_raw_parts.
+        assert!(!data.is_null());
+        assert!(isize::try_from(len).is_ok());
+
+        let slice =
+            unsafe { core::slice::from_raw_parts(data as *const (), len) };
+        slice as *const [()] as *const _
+    }
+
+    pub fn into_raw(arc: Arc<T>) -> *const T {
+        let ptr = unsafe { &(*arc.ptr).inner };
+        #[allow(clippy::mem_forget)]
+        mem::forget(arc);
+        ptr
+    }
+
+    pub unsafe fn from_raw(ptr: *const T) -> Arc<T> {
+        let align =
+            std::cmp::max(mem::align_of::<T>(), mem::align_of::<AtomicUsize>());
+
+        let rc_width = std::cmp::max(align, mem::size_of::<AtomicUsize>());
+
+        let sub_ptr = (ptr as *const u8).sub(rc_width) as *mut ArcInner<T>;
+
+        Arc { ptr: sub_ptr }
+    }
+}
+
+impl<T: ?Sized> Arc<T> {
+    pub fn strong_count(arc: &Arc<T>) -> usize {
+        unsafe { (*arc.ptr).rc.load(Ordering::Acquire) }
+    }
+
+    pub fn get_mut(arc: &mut Arc<T>) -> Option<&mut T> {
+        if Arc::strong_count(arc) == 1 {
+            Some(unsafe { &mut arc.ptr.as_mut().unwrap().inner })
+        } else {
+            None
+        }
+    }
+}
+
+impl<T: ?Sized + Clone> Arc<T> {
+    pub fn make_mut(arc: &mut Arc<T>) -> &mut T {
+        if Arc::strong_count(arc) != 1 {
+            *arc = Arc::new((**arc).clone());
+            assert_eq!(Arc::strong_count(arc), 1);
+        }
+        Arc::get_mut(arc).unwrap()
+    }
+}
+
+impl<T: Default> Default for Arc<T> {
+    fn default() -> Arc<T> {
+        Arc::new(T::default())
+    }
+}
+
+impl<T: ?Sized> Clone for Arc<T> {
+    fn clone(&self) -> Arc<T> {
```

Comment:
Cloning should increase the reference count. (currently rc only gets initialized, read, and decremented)

Reply:
not sure how that slipped my mind during the first pass hahaha